File: kafkaStreamsModelServerInMemoryStore/src/main/java/com/lightbend/java/naive/modelserver/DataProcessor.java
package com.lightbend.java.naive.modelserver;

import com.lightbend.model.Winerecord;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.java.model.DataConverter;
import com.lightbend.java.naive.modelserver.store.StoreState;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 * used
 * https://github.com/bbejeck/kafka-streams/blob/master/src/main/java/bbejeck/processor/stocks/StockSummaryProcessor.java
 */
public class DataProcessor extends AbstractProcessor<byte[], byte[]> {

    private StoreState modelStore;

    @Override
    public void process(byte[] key, byte[] value) {
        Optional<Winerecord.WineRecord> dataRecord = DataConverter.convertData(value);
        if(!dataRecord.isPresent()) {
            return;                                 // Bad record
        }
        if(modelStore.getNewModel() != null){
            // update the model
            if(modelStore.getCurrentModel() != null)
                modelStore.getCurrentModel().cleanup();
            modelStore.setCurrentModel(modelStore.getNewModel());
            modelStore.setCurrentServingInfo(new ModelServingInfo(modelStore.getNewServingInfo().getName(),
                    modelStore.getNewServingInfo().getDescription(), System.currentTimeMillis()));
            modelStore.setNewServingInfo(null);
            modelStore.setNewModel(null);
        }
        // Actually score
        if(modelStore.getCurrentModel() == null) {
            // No model currently
            System.out.println("No model available - skipping");
        }
        else{
            // Score the model
            long start = System.currentTimeMillis();
            double quality = (double) modelStore.getCurrentModel().score(dataRecord.get());
            long duration = System.currentTimeMillis() - start;
            modelStore.getCurrentServingInfo().update(duration);
            System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
         }

    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = StoreState.getInstance();

    }
}

File: kafkaStreamsModelServerInMemoryStore/src/main/scala/com/lightbend/scala/naive/modelserver/DataProcessor.scala
package com.lightbend.scala.naive.modelserver

import com.lightbend.scala.modelServer.model.{DataRecord, ServingResult}
import com.lightbend.scala.naive.modelserver.store.StoreState
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext}

import scala.util.Success

class DataProcessor extends AbstractProcessor[Array[Byte], Array[Byte]]{

  private var modelStore = null.asInstanceOf[StoreState]
  private var ctx = null.asInstanceOf[ProcessorContext]

  override def process(key: Array[Byte], value: Array[Byte]): Unit = {
    DataRecord.fromByteArray(value) match {
      case Success(dataRecord) => {
        modelStore.newModel match {
          case Some(model) => {
            // close current model first
            modelStore.currentModel match {
              case Some(m) => m.cleanup()
              case _ =>
            }
            // Update model
            modelStore.currentModel = Some(model)
            modelStore.currentState = modelStore.newState
            modelStore.newModel = None
          }
          case _ =>
        }
        modelStore.currentModel match {
          case Some(model) => {
            val start = System.currentTimeMillis()
            val quality = model.score(dataRecord.asInstanceOf[AnyVal]).asInstanceOf[Double]
            val duration = System.currentTimeMillis() - start
//            println(s"Calculated quality - $quality calculated in $duration ms")
            modelStore.currentState.get.incrementUsage(duration)
            ctx.forward(key, ServingResult(true, quality, duration))
          }
          case _ => {
//            println("No model available - skipping")
            ctx.forward(key, ServingResult(false))
          }
        }
        ctx.commit()
      }
      case _ => // ignore
    }
  }

  override def init(context: ProcessorContext): Unit = {
    modelStore = StoreState()
    ctx = context
  }
}
